src/messaging/ReliableMessageMgr.cpp

```diff
@@ -106,138 +94,61 @@ void ReliableMessageMgr::TicklessDebugDumpRetransTable(const char * log)
-    // has expired
+    // Retransmit / cancel anything in the retrans table whose retrans timeout has expired
     mRetransTable.ForEachActiveObject([&](auto * entry) {
-        CHIP_ERROR err = CHIP_NO_ERROR;
-
-        if (entry->nextRetransTimeTick != 0)
+        if (entry->nextRetransTime > now)
             return true;
 
-        if (entry->retainedBuf.IsNull())
-        {
-            // We generally try to prevent entries with a null buffer being in a table, but it could happen
-            // if the message dispatch (which is supposed to fill in the buffer) fails to do so _and_ returns
-            // success (so its caller doesn't clear out the bogus table entry).
-            //
-            // If that were to happen, we would crash in the code below.  Guard against it, just in case.
-            ClearRetransTable(*entry);
-            return true;
-        }
+        VerifyOrDie(!entry->retainedBuf.IsNull());
 
         uint8_t sendCount       = entry->sendCount;
         uint32_t messageCounter = entry->retainedBuf.GetMessageCounter();
 
         if (sendCount == CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS)
         {
-            err = CHIP_ERROR_MESSAGE_NOT_ACKNOWLEDGED;
-
             ChipLogError(ExchangeManager,
                          "Failed to Send CHIP MessageCounter:" ChipLogFormatMessageCounter " on exchange " ChipLogFormatExchange
                          " sendCount: %" PRIu8 " max retries: %d",
                          messageCounter, ChipLogValueExchange(&entry->ec.Get()), sendCount, CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS);
 
-            // Remove from Table
-            ClearRetransTable(*entry);
+            // Do not StartTimer, we will schedule the timer at the final end
+            mRetransTable.ReleaseObject(entry);
+            return true;
         }
 
-        // Resend from Table (if the operation fails, the entry is cleared)
-        if (err == CHIP_NO_ERROR)
-            err = SendFromRetransTable(entry);
-
-        if (err == CHIP_NO_ERROR)
-        {
-            // If the retransmission was successful, update the passive timer
-            entry->nextRetransTimeTick = static_cast<uint16_t>(entry->ec->GetActiveRetransmitTimeoutTick());
-#if !defined(NDEBUG)
-            ChipLogDetail(ExchangeManager,
-                          "Retransmitted MessageCounter:" ChipLogFormatMessageCounter " on exchange " ChipLogFormatExchange
-                          " Send Cnt %d",
-                          messageCounter, ChipLogValueExchange(&entry->ec.Get()), entry->sendCount);
-#endif
-        }
+        ChipLogDetail(ExchangeManager,
+                      "Retransmitting MessageCounter:" ChipLogFormatMessageCounter " on exchange " ChipLogFormatExchange
+                      " Send Cnt %d",
+                      messageCounter, ChipLogValueExchange(&entry->ec.Get()), entry->sendCount);
+        entry->nextRetransTime = System::SystemClock().GetMonotonicTimestamp() + entry->ec->GetMRPConfig().mActiveRetransTimeout;
```

Comment:
Shouldn't we switch to the active timeout only if we received any ack on a given exchange? It may not have worked like that even before your change, but in such a case adding a TODO would be nice.

Reply:
TODO Added